crates/feldera-storage/src/backend/mod.rs

```diff
@@ -0,0 +1,244 @@
+pub enum ReadBuffer {
+    #[cfg(feature = "glommio")]
+    Glommio(ReadResult),
+    Vec(Vec<u8>),
+}
+
+impl Deref for ReadBuffer {
+    type Target = [u8];
+
+    fn deref(&self) -> &Self::Target {
+        match self {
+            #[cfg(feature = "glommio")]
+            ReadBuffer::Glommio(x) => x.as_ref(),
+            ReadBuffer::Vec(x) => x.as_slice(),
+        }
+    }
+}
+
+/// An error that can occur when using the storage backend.
+#[derive(Error, Debug)]
+pub enum StorageError {
+    #[cfg(feature = "glommio")]
+    #[error("Got IO error during glommio operation")]
+    Io(#[from] glommio::GlommioError<()>),
+    #[error("Got IO error during monoio operation")]
+    StdIo(#[from] std::io::Error),
+}
+
+/// A trait for a storage backend to implement so client can create/delete
+/// files.
+pub trait StorageControl {
+    /// Creates a new persistent file used for writing data.
+    ///
+    /// Returns a file-descriptor that can be used for writing data.
+    /// Note that it is not possible to read from this file until
+    /// [`StorageWrite::complete`] is called and the [`FileHandle`] is
+    /// converted to an [`ImmutableFileHandle`].
+    async fn create(&self) -> Result<FileHandle, StorageError>;
+
+    /// Deletes a previously completed file.
+    ///
+    /// This removes the file from the storage backend and makes it unavailable
+    /// for reading.
+    async fn delete(&self, fd: ImmutableFileHandle) -> Result<(), StorageError>;
+
+    /// Deletes a previously created file.
+    ///
+    /// This removes the file from the storage backend and makes it unavailable
+    /// for writing.
+    ///
+    /// Use [`delete`] for deleting a file that has been completed.
+    async fn delete_mut(&self, fd: FileHandle) -> Result<(), StorageError>;
+}
+
+/// A trait for a storage backend to implement so clients can write to files.
+pub trait StorageWrite {
+    /// Allocates a buffer suitable for writing to a file using Direct I/O over
+    /// `io_uring`.
+    fn allocate_buffer(sz: usize) -> WriteBuffer;
+
+    /// Writes a block of data to a file.
+    ///
+    /// ## Arguments
+    /// - `fd` is the file-handle to write to.
+    /// - `offset` is the offset in the file to write to.
+    /// - `data` is the data to write.
+    ///
+    /// ## Preconditions
+    /// - `data.len() >= 512`: Minimum block size for Direct I/O, otherwise the
+    ///   write will return an error.
+    /// - `data.len().is_power_of_two() == true`: The size of the buffer must be
+    ///   a power of two.
+    /// - `offset.is_power_of_two() == true`: The offset must be a power of two.
+    ///
+    /// ## Returns
+    /// The number of bytes written.
+    ///
+    /// API returns an error if any of the above preconditions are not met.
+    async fn write_block(
+        &self,
+        fd: &FileHandle,
+        offset: u64,
+        data: WriteBuffer,
+    ) -> Result<usize, StorageError>;
```

Comment:
It might be a good idea to return a ReadBuffer on the success side, since at least for our application it's likely that we're going to need what we wrote again soon. The buffer cache code could use it.

Reply:
Makes sense, maybe we can discuss implications offline.. e.g., currently the `complete()` call ensures that we fsync the whole file and return an `ImmutableFileHandle` before we can read from it and get ReadBuffers.

If we start to return ReadBuffers before the call to complete() I'm not sure if it has any downsides for making a robust API that prevents shooting ourselves in the foot with data-integrity later. (likely the answer can be it's fine, but I haven't thought about it too much yet)

Reply:
Hmm, I see what you mean. Maybe it is not worth this optimization initially.